### PR TITLE
feat: implement JIT component experiment

### LINKED ISSUE #29244: Add option for Component Testing to JIT compile test before executing.
### What would you like?

An option to Just-In-Time (JIT) compile CT tests as the test is being run instead of compiling all assets at once before suite execution. 

### Why is this needed?

As component testing suites grow, `cypress open` and `cypress run` performance tends to be noticeably slower. This is due to the webpack dev server / vite dev server compiling all assets in the `specPattern` on launch, which can be very slow if a user has a lot of component tests. This would make open mode (and run mode in some scenarios) faster when it comes to running a single test as waiting for a lengthy compile is no longer needed. 

### Other

Performance impacts are starting to be felt by the Cypress core team when developing component tests as well as some of our users who have larger component testing suites. 

Since this might add a configuration option, we may need to make the signature public to allow OSS developers of other CT dev servers the option to implement.

### PR DESCRIPTION
**Pull request by AtofStryker**
- Closes #29244

#### Some historical context: 

Earlier this year, users were and still are experiencing issues with `Chunk Load Error`s in #28644 where different aspects of the webpack bundle, hosted by `webpack-dev-server`, are timing out or failing to load. We believe this is due to high memory usage consumed by the dev server due to Cypress compiling all user assets at once needed to run every single component test. This means the dev servers scales with the size of the component testing project, which is not good ☹️ . Because of this, users have been running into the following:

- Out Of Memory (OOM) errors.
- `chunk load error`, specific to Webpack where parts of the bundle fail to load.
- long run times while running component tests in CI in `cypress run` mode.
- Unable to efficiently develop tests in `cypress open` mode because the initial and recompiling times for specs and related components is extremely long.
- Inability to leverage Smart orchestration effectively to scale resources due to all assets being compiled into the dev server regardless of what specs are run.

The proposed solution to deal with these problems is #29244, a Just-In-Time compilation option that only compiles the assets needed for a given spec right before the spec is run (hence the name just-in-time), with the idea reducing resources needed to run the dev server. That is exactly what this feature is, `experimentalJustInTimeCompile`.

#### experimentalJustInTimeCompile

Enabling `experimentalJustInTimeCompile` will only compile assets and resources related to the spec before the spec is run. This is applicable in both `open` and `run` modes.

In `open` and `run` modes, the dev server will start up when the Cypress application starts with the component testing type, but will only compile the spec once it is selected or run. When a different spec is run, the different spec is then compiled, omitting the previously run spec from the dev server.

This makes developing your current spec much faster in `open` mode when it comes to initial compiling and recompiling, with a minor tradeoff being to recompile when you switch specs.

This also significantly reduces the memory requirements needed to run your tests in `run` mode, with a tradeoff being that your tests might run slightly slower since the dev server needs to recompile. If time is a factor, Smart orchestration can now be effectively used to parallelize your component testing runs on smaller infrastructure since the dev server only compiles what is needed to run the spec.

### Steps to test

Since we need a large project to test against, most of the benchmarking was done against internal Cypress repositories. From our experimentation, we saw memory usage drop about 40% with run times being close to the same maybe a little longer for `run` mode. For `open` mode, we saw initial compile times reduced by about 60%.

If testing in the cypress repository, there are system tests and cypress-in-cypress tests added to verify both run and open mode. The system tests make sure the dev server is only started once, but the entries are updated as the run progresses through the specs. All specs should be passing for both `webpack` and `vite`. The cy-in-cy tests verify that the user can switch between specs in the Cypress app and that the specs pass. If you have the debug flags enabled for `cypress:webpack*` or `cypress:vite*`, you will see the entry update to the spec that is selected.

### How has the user experience changed?
Users can hopefully leverage this experiment to unburden themselves of the above errors when running their tests in CI and increasing developer productivity with `open` mode via not having to wait significantly long periods of time for tests to compile.

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/5897
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: system-tests/__snapshots__/results_spec.ts.js [hunk 1]
```diff
@@ -27,6 +27,7 @@ exports['module api and after:run results'] = `
     "experimentalMemoryManagement": false,
     "experimentalModifyObstructiveThirdPartyCode": false,
     "experimentalSkipDomainInjection": null,
+    "experimentalJustInTimeCompile": false,
```
Comment (jennifer-shehane) [inline system-tests/__snapshots__/results_spec.ts.js:30]: I approve this change to this results_spec.ts.js file, as a code owner.

File: cli/types/cypress.d.ts [hunk 2]
```diff
@@ -3124,6 +3124,14 @@ declare namespace Cypress {
      * @default null
      */
     experimentalSkipDomainInjection: string[] | null
+    /**
+     * Allows for just-in-time compiling of a component test, which will only compile assets related to the component.
+     * This results in a smaller bundler under test, reducing resource constraints on a given machine. This option is recommended
```
Comment (ryanthemanuel) [inline cli/types/cypress.d.ts]:
```suggestion
     * This results in a smaller bundle under test, reducing resource constraints on a given machine. This option is recommended
```

File: system-tests/projects/experimental-JIT/vite/cypress/support/component-index.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  <head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width,initial-scale=1.0">
+    <title>Components App</title>
+  </head>
+  <body>
+    <div data-cy-root></div>
+  </body>
+</html>
```
Comment (ryanthemanuel) [inline system-tests/projects/experimental-JIT/vite/cypress/support/component-index.html]:
```suggestion
</html>

```

File: system-tests/projects/experimental-JIT/webpack/cypress/support/component-index.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  <head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width,initial-scale=1.0">
+    <title>Components App</title>
+  </head>
+  <body>
+    <div data-cy-root></div>
+  </body>
+</html>
```
Comment (ryanthemanuel) [inline system-tests/projects/experimental-JIT/webpack/cypress/support/component-index.html]:
```suggestion
</html>

```

File: npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts [hunk 5]
```diff
@@ -98,7 +99,8 @@ export function makeCypressWebpackConfig (
     devtool: 'inline-source-map',
   } as any
 
-  if (isRunMode) {
+  // if experimentalJustInTimeCompile is configured, we need to watch for file changes as the spec entries are going to be updated per test
```
Comment (ryanthemanuel) [inline npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts]: Do we need to watch everything? Or is there a smaller subset of files that we can watch in the JIT case?

Comment (AtofStryker) [reply npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts]: It's actually doing the later. The watching only applies to files that are in the entry, which is the support file, index file and related files, plus the spec entry. I contemplated narrowing this down to just the specpattern, but we have this weird [hack](https://github.com/cypress-io/cypress/blob/develop/npm/webpack-dev-server/src/CypressCTWebpackPlugin.ts#L122) where we save the index file with no changes to prompt the dev server to trigger a refresh. I wonder if we can just do the index html file and we get the same effect? Ill try it

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 6]
```diff
@@ -252,13 +260,13 @@ export class ProjectLifecycleManager {
...
 
-          span?.end()
-
           if (!devServerOptions?.port) {
             throw getError('CONFIG_FILE_DEV_SERVER_INVALID_RETURN', devServerOptions)
           }
 
           finalConfig.baseUrl = `http://localhost:${devServerOptions?.port}`
+
+          span?.end()
```
Comment (cacieprins) [inline packages/data-context/src/data/ProjectLifecycleManager.ts:270]: if line 264 gets hit, this span doesn't end - what effect does an orphan span have on the telemetry for `dataContext:ct:startDevServer` ?

**Review by cacieprins — APPROVED** [review]

File: packages/data-context/src/data/ProjectLifecycleManager.ts (hunk 6 above)
Comment (AtofStryker) [reply packages/data-context/src/data/ProjectLifecycleManager.ts:270]: the spans look super messy. Ill fix this. Thank you for catching this!

Comment (AtofStryker) [reply packages/data-context/src/data/ProjectLifecycleManager.ts:270]: addressed in e4eaaf8986

File: npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts (hunk 5 above)
Comment (AtofStryker) [reply npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts]: @ryanthemanuel So the watch options for webpack are a bit complicated, but I figure a quick win here is to at least ignore the `node_modules` directory. Vite ignores this by [default](https://vitejs.dev/config/server-options#server-watch). updated watch options in 77b5dee6da

**Review by ryanthemanuel — APPROVED** [review]

**Review by jennifer-shehane — APPROVED** [review]

**Pull request merged**